### DISCUSSION TITLE
Possibility to force retain

### DISCUSSION THREAD

**Discussion by bartbutenaers** [moscajs/aedes#941] [Q&A] [answered]
Hi all,

I was wondering if there is a way to force retain for all topics?  Similar to the Mosquitto [force-retain](https://github.com/eclipse/mosquitto/tree/develop/plugins/examples/force-retain) plugin.

Because it seems that Shelly gen 2 devices (based on esp32) [don't](https://www.reddit.com/r/shellycloud/comments/197bs6p/mqtt_on_gen2/) offer a way to activate the retain flag in their web interfaces.  With their gen 1 devices (based on esp8266) that feature is available.  So all these messages from my new Shelly devices unfortunately arrive in my Aedes broker as retain false, which is not sufficient for my use case.

Although I understand that this is not a standard mqtt feature, it would be very useful for such use cases.  Or if you could apply this setting to specific topics only, that would even be better...

Or if somebody has another workaround, all tips are appreciated!

Thanks!
Bart

**Comment by gnought** [comment]
@bartbutenaers it is a good suggestion but also a risk.

afaik currently aedes does not support modify PUBLISH packet on the fly, this also open a risk to misbehave the core when a modified packet is malformed, which require developers to take care of.

**Comment by robertsLando** [comment, marked as answer]
@bartbutenaers I think you can already edit the packet in  the `authorizePublish` hook but be aware that you may have some unwanted results

**Comment by bartbutenaers** [comment]
Hi @gnought, @robertsLando,
That `authorizePublish` hook might indeed do the job for me.  Thanks for the tip!!
You are talking about *"risk"* and *"unwanted results"*.  Could you please explain that a bit more in detail please, because I am not really an MQTT expert.  My use case is home automation so the data traffic is quite low, in case you are reffering to that.

**Comment by bartbutenaers** [comment]
I think I am doing something wrong.  When I add this handler:
```
    const broker = aedes.createBroker(aedesSettings);

    broker.authorizePublish = (client, packet, callback) => {
      if (packet.topic.startsWith('shelly')) {
        packet.retain = true
      }

      callback(null)
    }
```
Then indeed this code will set the retain flag to true, when a message from my shelly arrives:

![image](https://github.com/moscajs/aedes/assets/14224149/1b128101-ed58-4d86-8fd2-a85c0aed28c8)

However in my mqtt client application (Node-RED) it looks like the arriving message has still the retain flag as false:

![image](https://github.com/moscajs/aedes/assets/14224149/4658ba7b-9f92-4f3b-a776-af75bb7e96dc)

Do I perhaps need to add some extra code to update the value?

**Reply by robertsLando** [reply]
@bartbutenaers but when the client connect is that message sent on subscription?

**Reply by bartbutenaers** [reply]
@robertsLando,
Back home, so did the test again...

Now I see that the message arrives indeed after my client reconnect, and I also see that the retain flag is `true`:

![image](https://github.com/moscajs/aedes/assets/14224149/0dc6f05a-d636-472d-ba74-816266f70e97)

I assume I somehow tested it incorrectly last night.
Sorry for the noise and thank you very much for all the help!!!

Case closed ;-)

Bart
